[Profiler][Minor] Set end time on python events when profiling stops.

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83622
* __->__ #83621
* #83033
* #82993
* #82992

We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

Commit:
[Profiler][Minor] Set end time on python events when profiling stops.

We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

Commit:
Update on "[Profiler][Minor] Set end time on python events when profiling stops."


We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

slgong-fb reviewed (approved):
Looks good to me! Need to take care of CI tests that are currently failing.

Commit:
Update on "[Profiler][Minor] Set end time on python events when profiling stops."


We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

robieta commented:
> Looks good to me! Need to take care of CI tests that are currently failing.

I think those are from an issue with an earlier PR, but we'll see. I also realized I should only be adding `SOFT_ASSERT`s, not any more `TORCH_INTERNAL_ASSERT`s.

Commit:
Update on "[Profiler][Minor] Set end time on python events when profiling stops."


We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

Commit:
Update on "[Profiler][Minor] Set end time on python events when profiling stops."


We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

Commit:
Update on "[Profiler][Minor] Set end time on python events when profiling stops."


We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

Commit:
Update on "[Profiler][Minor] Set end time on python events when profiling stops."


We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

Commit:
Update on "[Profiler][Minor] Set end time on python events when profiling stops."


We don't have an end event for calls that are ongoing when profiling stops. (e.g. main) This cropped up when I was adding checks for negative durations.

I also refactored `populate` to use a pop method. This not only allows me to implement this fix, but should also provide a convenient entry point for https://github.com/pytorch/pytorch/pull/82154

Differential Revision: [D38426342](https://our.internmc.facebook.com/intern/diff/D38426342/)

[ghstack-poisoned]

robieta commented:
@pytorchbot merge -l

robieta commented:
@pytorchbot merge

robieta commented:
@pytorchbot merge -f "I see nothing in the logs for the failed clang job to suggest that this stack caused it to fail. It seems to have just hung."

zejun-chen commented on torch/csrc/profiler/collection.cpp:
```diff
@@ -447,13 +448,21 @@ uint64_t Result::correlationID() const {
       ATTRIBUTE(Kineto, start_time_ns_ + e.duration_us_ * 1000),
       [&](const auto& e) -> int64_t { return e.end_time_ns_; }));
+
+  // In rare cases we're willing to tolerate ops which are missing an end time
```
Hi @robieta 
May i know what the rare cases are?
